### PR TITLE
Print deprecation message when implicit block expectation syntax is used

### PR DESCRIPTION
**Pull request by pirj**
The purpose of this (rather controversial) change is to discourage the usage of a so-called implicit block expectation syntax:

```ruby
subject { -> { do_something } }
it { is_expected.to change(something).to(new_value) }
```

In case this syntax is being used, a deprecation warning will be printed.

This syntax was referenced to as:

> pretty obtuse and not something I’d recommend, generally.
https://github.com/rspec/rspec-expectations/issues/805#issuecomment-112239820

> We should mention they’re not supported.
https://github.com/rspec/rspec-expectations/issues/805#issuecomment-112238533

From [`rspec-core` source](https://github.com/rspec/rspec-core/blob/725b0d2eadaa9ac9f816a6d2d7ef1e507a6705cd/features/subject/one_liner_syntax.feature#L25):

> The one-liner syntax only works with non-block expectations (e.g. `expect(obj).to eq`, etc) and it cannot be used with block expectations (e.g. `expect { object }`).

This is a less strict follow-up to #1125, which was released as part of a patch release 3.8.5, and later reverted in 3.8.6 due to a number of complaints.
Even though [workarounds are possible for the most desperate usages of this syntax](https://github.com/rodjek/rspec-puppet/pull/767), introducing a deprecation warning in pre-4.0 version of RSpec and removal of (coincidental, but still)  support for the implicit block syntax in a major release makes more sense than removal of it in a minor/patch version, since it's used quite often across different codebases I've seen.

### Example Deprecation Message

```
Deprecation Warnings:

You must pass a block rather than an argument to `expect` to use the provided block expectation matcher (change `1`), or the matcher must implement `supports_value_expectations?`.
```

I'm not quite certain it's the best one possible, since it does not refer to the code that led to this deprecation message.
Appreciate any hint how to make it better.

### Additional Things to Notice

According to https://github.com/rspec/rspec-expectations/pull/1125#issuecomment-539429318, the following works with no disruptions:

```ruby
RSpec.describe 'introspect' do
  it do
    expect(-> {}).to be_a(Proc)
  end
end
```

### Related Links

https://github.com/rspec/rspec-expectations/issues/526
https://github.com/rspec/rspec-expectations/pull/530
https://github.com/rspec/rspec-expectations/issues/536
https://github.com/rspec/rspec-expectations/pull/1066
https://github.com/rspec/rspec-expectations/pull/1125

https://blog.rubystyle.guide/rspec/2019/07/17/rspec-implicit-block-syntax.html
https://www.reddit.com/r/ruby/comments/cejl3q/call_for_discussion_rspec_implicit_block
https://github.com/rubocop-hq/rspec-style-guide/issues/76
https://github.com/rubocop-hq/rubocop-rspec/pull/789

Example fix for matcher libraries that are forced to keep the interface to provide the syntax for their users https://github.com/rodjek/rspec-puppet/pull/767

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -32,6 +32,11 @@ Bug Fixes:
 
 Bug Fixes:
 
+* Prevent unsupported implicit block expectation syntax from being used.
+  (Phil Pirozhkov, #1125)
+
+Bug Fixes:
+
```
Comment (JonRowe) [inline Changelog.md]: This needs sorting out, its not right...

File: lib/rspec/expectations/expectation_target.rb [hunk 2]
```diff
@@ -118,9 +157,11 @@ def enforce_block_expectation(matcher)
...
 
       def supports_block_expectations?(matcher)
-        matcher.supports_block_expectations?
-      rescue NoMethodError
-        false
+        if matcher.respond_to?(:supports_block_expectations?)
+          matcher.supports_block_expectations?
+        else
+          false
+        end
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb]: This can go back as it was

File: Changelog.md [hunk 3]
```diff
@@ -51,7 +56,7 @@ Bug Fixes:
 * Prevent composed `all` matchers from leaking into their siblings leading to duplicate
   failures. (Jamie English, #1086)
 * Prevent objects which change their hash on comparison from failing change checks.
-  (Phil Pirozhkov, #1110)
+  (Phil Pirozhkov, #1100)
```
Comment (JonRowe) [inline Changelog.md]: Is this right?

File: lib/rspec/expectations/expectation_target.rb [hunk 4]
```diff
@@ -33,16 +33,16 @@ def initialize(value)
...
-      def self.for(value, block)
+      def self.for(value, &block)
         if UndefinedValue.equal?(value)
-          unless block
+          unless block_given?
             raise ArgumentError, "You must pass either an argument or a block to `expect`."
           end
           BlockExpectationTarget.new(block)
-        elsif block
+        elsif block_given?
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb]: Why is the above changing?

File: lib/rspec/expectations/expectation_target.rb [hunk 5]
```diff
@@ -90,6 +90,45 @@ def prevent_operator_matchers(verb)
...
+
+    private
+
+      def enforce_value_expectation(matcher)
+        return if supports_value_expectations?(matcher)
+
+        RSpec.deprecate(
+          "expect(value).to #{RSpec::Support::ObjectFormatter.format(matcher)}",
+          :message =>
+            "You must pass a block rather than an argument to `expect` to use the provided " \
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb]: Does this message need rewording? It should probably talk about it being recommended rather than must.

File: lib/rspec/expectations/expectation_target.rb [hunk 6]
```diff
@@ -90,6 +90,45 @@ def prevent_operator_matchers(verb)
...
+            "or the matcher must implement `supports_value_expectations?`."
+        )
+      end
+
+      def supports_value_expectations?(matcher)
+        if matcher.respond_to?(:supports_value_expectations?)
+          matcher.supports_value_expectations?
+        else
+          true
+        end
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb]: As per below, call the method, rescue the `NoMethodError` its better ruby

File: lib/rspec/matchers/built_in/change.rb [hunk 7]
```diff
@@ -242,11 +253,6 @@ def did_change_failure
           "did change from #{@actual_before_description} " \
           "to #{description_of @change_details.actual_after}"
         end
-
-        def not_given_a_block_failure
-          "expected #{@change_details.value_representation} to have changed " \
-          "#{change_description}, but was not given a block"
-        end
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/change.rb]: Why are you removing this? Is this a revert artifact or ?

File: lib/rspec/matchers/built_in/change.rb [hunk 8]
```diff
@@ -337,6 +342,8 @@ def change_description
       class ChangeDetails
         attr_reader :actual_after
 
+        UNDEFINED = Module.new
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/change.rb]: Freeze this

File: lib/rspec/matchers/built_in/output.rb [hunk 9]
```diff
@@ -94,20 +94,25 @@ def supports_block_expectations?
...
+        end
+
       private
 
         def captured?
           @actual.length > 0
         end
 
         def positive_failure_reason
-          return "was not a block" unless Proc === @block
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/output.rb]: Why is this removed?

File: lib/rspec/matchers/built_in/raise_error.rb [hunk 10]
```diff
@@ -199,7 +205,6 @@ def format_backtrace(backtrace)
         end
 
         def given_error
-          return " but was not given a block" unless Proc === @given_proc
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/raise_error.rb]: Why is this removed?

File: lib/rspec/matchers/built_in/throw_symbol.rb [hunk 11]
```diff
@@ -88,20 +88,23 @@ def description
...
+
+        # @api private
         def expects_call_stack_jump?
           true
         end
 
       private
 
         def actual_result
-          return "but was not a block" unless Proc === @block
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/throw_symbol.rb]: Why is this removed?

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -5,8 +5,8 @@
   minimum_coverage 97
 end
 
-Dir['./spec/support/**/*'].each do |f|
-  require f.sub(%r{\./spec/}, '')
+Dir['./spec/support/**/*.rb'].each do |f|
+  require f.sub(%r|\./spec/|, '')
```
Comment (JonRowe) [inline spec/spec_helper.rb]: Revert this unless needed please :)

File: lib/rspec/matchers/built_in/yield.rb [hunk 13]
```diff
@@ -405,23 +404,24 @@ def supports_block_expectations?
...
         def positive_failure_reason
-          return 'was not a block' unless @probe.has_block?
-
           'yielded with unexpected arguments' \
           "\nexpected: #{surface_descriptions_in(@expected).inspect}" \
           "\n     got: [#{@actual_formatted.join(", ")}]"
         end
 
         def negative_failure_reason
-          return 'was not a block' unless @probe.has_block?
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/yield.rb]: Why is this removed?

File: lib/rspec/matchers/built_in/yield.rb [hunk 14]
```diff
@@ -405,23 +404,24 @@ def supports_block_expectations?
...
+        end
+
       private
 
         def expected_arg_description
           @expected.map { |e| description_of e }.join(', ')
         end
 
         def positive_failure_reason
-          return 'was not a block' unless @probe.has_block?
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/yield.rb]: Why is this removed?

File: lib/rspec/matchers/built_in/yield.rb [hunk 15]
```diff
@@ -10,7 +10,6 @@ module BuiltIn
       class YieldProbe
         def self.probe(block, &callback)
           probe = new(block, &callback)
-          return probe unless probe.has_block?
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/yield.rb]: I'm not sure I understand any of this files changes... can you elaborate?

**Review by JonRowe — DISMISSED** [review]
Sorry for the delay @pirj got a bit exhausted by this change then forgot about!

**Comment by pirj** [comment]
@JonRowe Is it better finishing this before we release 4.0? Or do you think it will be ok to introduce this warning in 4.0.x and remove in 4.1.0?

**Comment by JonRowe** [comment]
> Or do you think it will be ok to introduce this warning in 4.0.x and remove in 4.1.0?
No that would be a breaking change, any removal would be in a major version.

**Comment by pirj** [comment]
Sounds good 👍 
I intend to wrap this PR up for 3.11 or 3.12, and handle the removal for 4.0.

File: spec/spec_helper.rb (hunk 12 above)
Comment (pirj) [reply spec/spec_helper.rb:9]: I can't:
```
LoadError:
  cannot load such file -- support/shared_examples
```

Comment (pirj) [reply spec/spec_helper.rb]: Reverted the `%r{}` -> `%r||` change. It triggers RuboCop locally, but we don't check `spec/` on CI.

**Comment by pirj** [comment]
The one and only diff-lcs example failure on Travis seems unrelated.

File: lib/rspec/expectations/expectation_target.rb (hunk 5 above)
Comment (JonRowe) [reply lib/rspec/expectations/expectation_target.rb:117]: > I don't recall how it was done in 2.99, don't we fail on deprecation warnings in the pre-major version?
No. Or they wouldn't be warnings.

Comment (JonRowe) [reply lib/rspec/expectations/expectation_target.rb:117]: Rather than `"You must"` rephrase to `"<x> is deprecated, and you should do <y> instead"`

File: spec/support/shared_examples/block_matcher.rb [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+    unless options[:disallows_negation]
+      expect {
+        expect(3).not_to matcher
+      }.to fail_with(/was not( given)? a block/)
+    end
+  end
+
+  it 'prints a deprecation warning when given a value' do
+    expect_warn_deprecation(/You must pass a block/)
```
Comment (pirj) [inline spec/support/shared_examples/block_matcher.rb:82]: Spotted https://github.com/rspec/rspec-expectations/issues/1233

File: lib/rspec/expectations/expectation_target.rb (hunk 5 above)
Comment (pirj) [reply lib/rspec/expectations/expectation_target.rb]: The message is fixed.

**Comment by pirj** [comment]
Hold on, I used 1.9 hash syntax, let me fix it real quick.

**Comment by JonRowe** [comment]
I'll try to review this at the weekend, its too large to do during this week with my work schedule atm.

**Comment by pirj** [comment]
Ping just in case.

**Comment by pirj** [comment]
Ping.

**Comment by JonRowe** [comment]
@pirj this is still on my to do list don't worry, it's just lower than rspec-rails 6.1 support (please don't send ping comments)

File: lib/rspec/expectations/expectation_target.rb [hunk 17]
```diff
@@ -42,7 +42,7 @@ def self.for(value, block)
         elsif block
           raise ArgumentError, "You cannot pass both an argument and a block to `expect`."
         else
-          new(value)
+          ValueExpectationTarget.new(value)
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb:45]: Do we now ever instantiate an `ExpectationTarget`? Could/should this class be merged into it?

Comment (JonRowe) [reply lib/rspec/expectations/expectation_target.rb:45]: Oh wait it's because `BlockExpectationTarget` inherits from it too right?

**Review by JonRowe — COMMENTED** [review]
I've mostly reviewed this, just need to do some sanity checking on the shared examples when I get a bit more time.

What I am a little concerned about is the proximity to release. We won't get a lot of feedback.

**Comment by pirj** [comment]
We had some feedback in https://github.com/rspec/rspec-expectations/issues/1134
I've ["fixed" it for `rspec-puppet`](https://github.com/rodjek/rspec-puppet/pull/767/files#diff-745b723dc6f8291ba9a7cc938d38bde2R13) including a [monkey-patched `raise_error` matcher](https://github.com/rodjek/rspec-puppet/pull/767/files#diff-225bb60162eb52e6e6555a13d9f882ad5dedbe6fe987a51dabb4c302c882b6deR13).

My idea was to:
 - deprecate in 3.x (if something goes wrong - we can fix in 3-9-maintenance)
 - remove completely (basically re-apply #1125) in 4.0

**Comment by pirj** [comment]
I've just realized that one-liner `should`/`should_not` from Core  `enforce_value_expectation` check as they instantiate `PositiveExpectationHandler`/`NegativeExpectationHandler` directly. So,

```ruby
subject(:action) { -> { raise } }
it { should raise_error StandardError }
```
will **not** print a warning, while when `should` is replaced with `is_expected.to`, it will.

Green light to change those two definitions as below?
```diff
      def should(matcher=nil, message=nil)
-        RSpec::Expectations::PositiveExpectationHandler.handle_matcher(subject, matcher, message)
+        expect(subject).to(matcher, message)
      end

      def should_not(matcher=nil, message=nil)
-        RSpec::Expectations::NegativeExpectationHandler.handle_matcher(subject, matcher, message)
+        expect(subject).not_to(matcher, message)
      end
```
Is there a good reason it wasn't like that in the first place?

**Comment by pirj** [comment]
@JonRowe I'd love to work on the removal of this syntax in `4-0-dev` sometimes maybe Sunday or Tuesday, and the idea was to try to merge `main` in `4-0-dev`. So a gentle reminder.

**Comment by pirj** [comment]
Rebased and resolved a conflict.

**Comment by pirj** [comment]
I intend to incorporate changes to specs made here and add an error for implicit block expectation syntax to `4-0-dev`.
@JonRowe Do you have major concerns regarding this change?

File: lib/rspec/expectations/expectation_target.rb [hunk 18]
```diff
@@ -90,6 +90,44 @@ def prevent_operator_matchers(verb)
...
+      def enforce_value_expectation(matcher)
+        return if supports_value_expectations?(matcher)
+
+        RSpec.deprecate(
+          "expect(value).to #{RSpec::Support::ObjectFormatter.format(matcher)}",
+          :message =>
+            "The implicit block expectation syntax is deprecated, you should pass " \
+            "a block rather than an argument to `expect` to use the provided " \
+            "block expectation matcher (#{RSpec::Support::ObjectFormatter.format(matcher)}), " \
+            "or the matcher must implement `supports_value_expectations?`."
```
Comment (JonRowe) [inline lib/rspec/expectations/expectation_target.rb:120]:
```suggestion
            "The implicit block expectation syntax is deprecated, you should pass " \
            "a block rather than an argument to `expect` to use the provided " \
            "block expectation matcher or the matcher must implement " \
            "`supports_value_expectations?`. e.g  `expect { value }.to " \
            "#{RSpec::Support::ObjectFormatter.format(matcher)}` not " \
            "`expect(value).to #{RSpec::Support::ObjectFormatter.format(matcher)}`
```

**Comment by JonRowe** [comment]
Ok! I have finally reviewed this. I have the following concerns:

* That this change is going to be unwelcome based on previous feedback.

I think I'm happy to proceed regardless however, as we can always walk it back if we get enough feedback, and most of these changes are good anyway.

* That there is no example of the deprecated code in the deprecation.

e.g. 

> The implicit block expectation syntax is deprecated, you should pass a block rather than an argument to `expect` to use the provided block expectation matcher (raise Exception), or the matcher must implement `supports_value_expectations?`.

I would prefer something like the suggestion I've applied.

**Comment by pirj** [comment]
Sounds good. Applied the suggestion here, and made a similar change in #1285

**Comment by JonRowe** [comment]
Thanks for your patience and hard work on this :)
